File: agent-adapter/python/turboflow_adapter/backends/aider.py
"""Aider backend — wraps the `aider` CLI."""

from __future__ import annotations

import os

from turboflow_adapter.backend import AgentBackend
from turboflow_adapter.process import run, command_exists
from turboflow_adapter.types import (
    BackendCapabilities,
    ExecOptions,
    ExecResult,
    HealthCheckResult,
)


class AiderBackend(AgentBackend):
    def __init__(self) -> None:
        super().__init__("aider")

    @property
    def name(self) -> str:
        return "Aider"

    @property
    def description(self) -> str:
        return "AI pair programming CLI with multi-model support (Apache 2.0)"

    @property
    def url(self) -> str:
        return "https://aider.chat"

    @property
    def license(self) -> str:
        return "Apache-2.0"

    @property
    def capabilities(self) -> BackendCapabilities:
        return BackendCapabilities(
            multi_model=True,
            interactive=True,
            headless=True,
            streaming=True,
            bedrock=True,
            worktrees=True,
        )

    def resolve_model(self, model: str) -> str:
        is_bedrock = os.environ.get("CLAUDE_CODE_USE_BEDROCK") == "1"
        tier_map = {
            "opus": (
                f"bedrock/{os.environ.get('ANTHROPIC_DEFAULT_OPUS_MODEL', 'us.anthropic.claude-opus-4-6-v1')}",
                "claude-3-opus-20240229",
            ),
            "sonnet": (
                f"bedrock/{os.environ.get('ANTHROPIC_DEFAULT_SONNET_MODEL', 'us.anthropic.claude-sonnet-4-6')}",
                "claude-sonnet-4-20250514",
            ),
            "haiku": (
                f"bedrock/{os.environ.get('ANTHROPIC_DEFAULT_HAIKU_MODEL', 'us.anthropic.claude-haiku-4-5-20251001-v1:0')}",
                "claude-haiku-4-5-20251001",
            ),
        }
        tier = tier_map.get(model)
        if tier:
            return tier[0] if is_bedrock else tier[1]
        return model

    async def exec(self, prompt: str, options: ExecOptions | None = None) -> ExecResult:
        opts = options or ExecOptions()
        args: list[str] = []

        if opts.model:
            args.extend(["--model", self.resolve_model(opts.model)])
        if opts.files:
            args.extend(opts.files)
        if opts.headless or opts.print_only:
            args.extend(["--message", prompt, "--no-auto-commits", "--yes"])
        args.extend(opts.extra_args or [])

        return run(
            "aider", args=args, cwd=opts.cwd, env=opts.env, timeout=opts.timeout, stream=True
        )

    async def version(self) -> str | None:
        result = run("aider", args=["--version"], timeout=10)
        return result.stdout.strip() if result.exit_code == 0 else None

    async def is_installed(self) -> bool:
        return command_exists("aider")

    async def health_check(self) -> HealthCheckResult:
        installed = await self.is_installed()
        ver = await self.version() if installed else None
        warnings: list[str] = []
        details: dict[str, str | bool] = {}

        providers = []
        if os.environ.get("ANTHROPIC_API_KEY"):
            providers.append("Anthropic")
        if os.environ.get("OPENAI_API_KEY"):
            providers.append("OpenAI")
        if os.environ.get("CLAUDE_CODE_USE_BEDROCK") == "1":
            providers.append("Bedrock")
        if not providers:
            warnings.append("No model provider configured")

        details["providers"] = ", ".join(providers) or "none"
        return HealthCheckResult(
            installed=installed,
            version=ver,
            provider=providers[0] if providers else None,
            details=details,
            warnings=warnings,
        )

    async def install(self) -> None:
        self.log.info("Installing Aider...")
        if command_exists("pipx"):
            result = run("pipx", args=["install", "aider-chat"], timeout=120, stream=True)
            if result.exit_code == 0:
                return
        pip = "pip3" if command_exists("pip3") else "pip"
        result = run(pip, args=["install", "--user", "aider-chat"], timeout=120, stream=True)
        if result.exit_code != 0:
            raise RuntimeError("Aider installation failed. Try: pipx install aider-chat")

File: agent-adapter/ts/src/backends/aider.ts
// =============================================================================
// TurboFlow Agent Adapter — Aider Backend
//
// Wraps the `aider` CLI (https://aider.chat).
// Multi-model support (OpenAI, Anthropic, Bedrock, local models).
// Does NOT support: MCP, Agent Teams.
// =============================================================================

import { AgentBackend } from '../backend.js';
import { run, commandExists } from '../process.js';
import type {
  BackendCapabilities,
  ExecOptions,
  ExecResult,
  HealthCheckResult,
} from '../types.js';

export class AiderBackend extends AgentBackend {
  readonly name = 'Aider';
  readonly description = 'AI pair programming CLI with multi-model support (Apache 2.0)';
  readonly url = 'https://aider.chat';
  readonly license = 'Apache-2.0';

  readonly capabilities: BackendCapabilities = {
    mcp: false,
    agentTeams: false,
    multiModel: true,
    interactive: true,
    headless: true,
    streaming: true,
    bedrock: true,
    worktrees: true,
  };

  constructor() {
    super('aider');
  }

  resolveModel(model: string): string {
    const isBedrock = process.env.CLAUDE_CODE_USE_BEDROCK === '1';

    const tierMap: Record<string, { bedrock: string; direct: string }> = {
      opus: {
        bedrock: `bedrock/${process.env.ANTHROPIC_DEFAULT_OPUS_MODEL ?? 'us.anthropic.claude-opus-4-6-v1'}`,
        direct: 'claude-3-opus-20240229',
      },
      sonnet: {
        bedrock: `bedrock/${process.env.ANTHROPIC_DEFAULT_SONNET_MODEL ?? 'us.anthropic.claude-sonnet-4-6'}`,
        direct: 'claude-sonnet-4-20250514',
      },
      haiku: {
        bedrock: `bedrock/${process.env.ANTHROPIC_DEFAULT_HAIKU_MODEL ?? 'us.anthropic.claude-haiku-4-5-20251001-v1:0'}`,
        direct: 'claude-haiku-4-5-20251001',
      },
    };

    const tier = tierMap[model];
    if (tier) {
      return isBedrock ? tier.bedrock : tier.direct;
    }
    return model;
  }

  async exec(prompt: string, options: ExecOptions = {}): Promise<ExecResult> {
    const args: string[] = [];

    if (options.model) {
      args.push('--model', this.resolveModel(options.model));
    }

    // Add target files
    if (options.files?.length) {
      args.push(...options.files);
    }

    if (options.headless || options.printOnly) {
      // Non-interactive: --message mode
      args.push('--message', prompt, '--no-auto-commits', '--yes');
    }

    args.push(...(options.extraArgs ?? []));

    return run({
      command: 'aider',
      args,
      cwd: options.cwd,
      env: options.env,
      timeout: options.timeout,
      stream: true,
    });
  }

  async version(): Promise<string | null> {
    const result = await run({ command: 'aider', args: ['--version'], timeout: 10_000 });
    return result.exitCode === 0 ? result.stdout.trim() : null;
  }

  async isInstalled(): Promise<boolean> {
    return commandExists('aider');
  }

  async healthCheck(): Promise<HealthCheckResult> {
    const installed = await this.isInstalled();
    const version = installed ? await this.version() : null;
    const warnings: string[] = [];
    const details: Record<string, string | boolean> = {};

    // Check model providers
    const providers: string[] = [];
    if (process.env.ANTHROPIC_API_KEY) providers.push('Anthropic');
    if (process.env.CLAUDE_CODE_USE_BEDROCK === '1') providers.push('Bedrock');
    if (process.env.OPENAI_API_KEY) providers.push('OpenAI');
    if (process.env.OPENROUTER_API_KEY) providers.push('OpenRouter');

    if (providers.length === 0) {
      warnings.push('No model provider configured. Set ANTHROPIC_API_KEY, OPENAI_API_KEY, or OPENROUTER_API_KEY');
    }

    details.providers = providers.join(', ') || 'none';
    details.mcp = false;
    details.agentTeams = false;

    return {
      installed,
      version,
      provider: providers[0] ?? null,
      details,
      warnings,
    };
  }

  async install(): Promise<void> {
    this.log.info('Installing Aider...');

    // Try pipx first (isolated install)
    if (await commandExists('pipx')) {
      const result = await run({
        command: 'pipx',
        args: ['install', 'aider-chat'],
        timeout: 120_000,
        stream: true,
      });
      if (result.exitCode === 0) {
        this.log.info('Aider installed (pipx)');
        return;
      }
    }

    // Fallback to pip
    const pipCmd = await commandExists('pip3') ? 'pip3' : 'pip';
    const result = await run({
      command: pipCmd,
      args: ['install', '--user', 'aider-chat'],
      timeout: 120_000,
      stream: true,
    });

    if (result.exitCode === 0) {
      this.log.info('Aider installed (pip)');
      return;
    }

    throw new Error('Aider installation failed. Try: pipx install aider-chat');
  }
}
